### ISSUE TITLE
Add string words SEARCH and COMPARE

### ISSUE THREAD

**Issue by scotws** [scotws/TaliForth2#45]
Also consider **REPLACE, UNESCAPE**, and **SUBSTITUTE**, though they all are monsters in size.

**Comment by SamCoVT** [comment]
I'll give SEARCH and COMPARE a go.

**Comment by scotws** [comment]
Actually, there is already high-level Forth code for both words in the forth_code folder in the the file `temporary.fs`, I just completely forgot all about it. It's from pForth, that is, in the public domain. They are horribly long words, but once we have them in, we can add the tests, and then we can move them to assembler code. -- I'll let you do this, time is really getting to be a problem here :-( ...

**Comment by SamCoVT** [comment]
Interestingly enough, the definition for `SEARCH` didn't compile right away because it uses `U>`, which Tali doesn't have, to compare the lengths of the strings (Tali does have `U<` - apparently because I asked for it some time ago).  I've put in issue #98 and I'll take care of it.  For now, however, I've simply replaced `U>` with just `>`.  I don't even feel bad about doing this because Tali can't currently have strings > 32K due to memory restrictions.  I'll probably get this finished up (with tests, and perhaps writing `U>`)this weekend and issue a pull request.

**Comment by SamCoVT** [comment]
I'm trying to add the tests for search and compare and I'm running into a couple of issues.

The first issue is that:
`s" abcd" s" " ( empty string ) compare --> true` however the version from pforth says false if the second string has a zero length:
```
: search   ( c-addr1 u1 c-addr2 u2 -- c-addr3 u3 f )
   rot 2dup                          ( copy lengths )
    ( This next line used to use u> for comparison of string lengths )
    ( but Tali doesn't have the word yet - SamCo 2018-07-06 )
   over swap >  swap 0=  or if      ( if u2>u1 or u2=0 )
      nip nip  false exit            ( exit with false flag )
   then
   -rot 2over                        ( save c-addr1 u1 )
   2swap tuck 2>r                    ( save c-addr2 u2 )
   - 1+ over +  swap                 ( make c-addr1 c-addr1+u1-u2 )
   2r> 2swap                         ( retrieve c-addr2 u2 )
   do
      2dup i over compare 0= if      ( if we find the string )
         2drop +  i tuck -           ( calculate c-addr3 u3 )
         true unloop exit            ( exit with true flag )
      then
   loop
   2drop false ;                     ( leave c-addr1 u1 false )
```

I may try rewriting pforth's version.  It's a big word, but the check for 0 is right near the beginning.

The second issue is SLITERAL. 
```
{ : s14 [ s1 ] sliteral ; -> } 
{ s1 s14 compare -> 0 }
{ s1 s14 rot = rot rot = -> -1 ( <true> ) 0 ( <false> ) }
```
The first line makes a copy of s1 named s14 (although tali just uses the existing s14 and adds the runtime pushes of the address and legth of the existing string).  The second line shows they have the same contents.  The third line appears to be checking that the lengths are the same (<true>) but that the addresses are different (the `<false>`, but Tali says `<true>`).

I know that our S" compiles things directly to the dictionary, and strings made with SLITERAL are not allowed to be modified, so at first I thought perhaps that we didn't need to actually make another copy.  The issue, though, would be with code like this:
```
pad 5 char a fill
: myaaaaa pad 5 sliteral ;
```
The issue is that sliteral would simply put the runtime in for the string in PAD, but PAD might be reused for something else.

I'm not fond of having sliteral make another copy of the string data in the dictionary, but that's what this test seems to be implying (unless I'm reading the test wrong).  The standard never explicitly says that the address added to the dictionary can't be the original address, but this test implies that.

I plan on fixing the SEARCH forth code, and leaving the SLITERAL test failing for now.  I'm interested in your thoughts on SLITERAL.  If we want to make this test pass, we will need to make SLITERAL make a copy of the string, and we will need to rewrite `S"` so it doesn't make two copies of the string (because it currently calls SLITERAL after copying the string data into the dictionary (when compiling)).

**Comment by scotws** [comment]
**COMPARE** -- I'm wondering if pforth is simply non-standard here, because the Gforth version returns `1` and not `-1` (true):
```
s" abcd" s" " compare . 1  ok
```
Though the specification of **COMPARE** gives me a headache (https://forth-standard.org/standard/string/COMPARE), this seems to be what is supposed to happen, because u1 is equal to u2 up to the length of u2 (duh with an empty string) and u2 is smaller than u1. So maybe we do need to rewrite this from scratch. I note that Gforth has **COMPARE** coded in assembler, which on the long term is probably the only way to slay that monster.

(Personally, I think **S=** as it is defined in the tests makes more sense, but who am I.)

As for **SLITERAL** -- you know the joke that analphabetism is caused by reading tests? -- the only way that the test would make sense is if the string is actually copied. Digging into Gforth seems to imply this as well: 
```
: SLiteral  
  tuck 2>r  POSTPONE AHEAD here 2r> mem, align >r  POSTPONE THEN r>  POSTPONE Literal  POSTPONE Literal ;
latestxt
interpret/compile: SLiteral
```
**AHEAD** cases the word to  jump over the other stuff to **THEN**, and the other stuff calls **MEM,** which looks like this:
```
: mem,  
  here over allot swap move ;
```
There's the **MOVE**. So it looks like, yes, we have to copy with **SLITERAL**, and yes, **S"** needs to be rewritten as well. Maybe at first, we want to revert to the official definition as provided in https://forth-standard.org/standard/string/SLITERAL
```
: S" ( "ccc<quote>" -- ) 
   [CHAR] " PARSE POSTPONE SLITERAL 
; IMMEDIATE
``` 
as a high-level word?
(I'm basically out for the next weeks now, will let you know ASAP when I'm back. Thanks for all the great stuff!)

**Comment by SamCoVT** [comment]
Both of those pForth words were non-ANS when it came to zero length strings, and the test cases showed it (yay for good tests!).  I'm kinda sorta proud of myself because I was actually able to read and follow the pForth definitions, see where they went wrong, and fix them.  Unfortunately I can't share that elation with most of my friends (in the same way I don't lend them my HP RPN calculator when they need one) but it was a nice moment nonetheless.

I wasn't able to revert to the official definition of S" as it would have left us back at 255 character strings (parse needs the input to fit in the input buffer), but I did rewrite SLITERAL and S".  I also made it so S" strings used in compile-mode don't get copied twice, so going ANS-compliant isn't quite as painful memory-wise.
